File: src/core/entity.ts
// Entity detection (words, line tips, boxes) and moves that keep attached
// lines connected. Ported from ASCIIFlow (client/draw/entity.ts), MIT © Lewis Hemens.
import { Box, boundingBox } from "./box";
import { ARROW_FOR, connections, connects, isArrow, isBoxDrawing, UNICODE } from "./glyphs";
import { isErase, Layer } from "./layer";
import { snap } from "./snap";
import { Direction, Vector } from "./vector";

const H = UNICODE.lineHorizontal;
const V = UNICODE.lineVertical;

export function isContent(value: string | null | undefined): value is string {
  return value != null && !isErase(value);
}

export function isText(value: string | null | undefined): value is string {
  return isContent(value) && !isBoxDrawing(value);
}

/** Maximal horizontal run of text characters under `p`. */
export function detectWord(layer: Layer, p: Vector): Vector[] | null {
  if (!isText(layer.get(p))) return null;
  let left = p;
  let right = p;
  while (isText(layer.get(left.left()))) left = left.left();
  while (isText(layer.get(right.right()))) right = right.right();
  const cells: Vector[] = [];
  for (let x = left.x; x <= right.x; x++) cells.push(new Vector(x, p.y));
  return cells;
}

export interface LineTip {
  tip: Vector;
  axis: "horizontal" | "vertical";
  /** Direction from the tip toward the rest of the line. */
  bodyDir: Direction;
  arrow: string | null;
}

/** The free end of a line or arrow under `p`, if any. */
export function detectLineTip(layer: Layer, p: Vector): LineTip | null {
  const value = layer.get(p);
  let axis: LineTip["axis"];
  if (value === H) axis = "horizontal";
  else if (value === V) axis = "vertical";
  else if (isArrow(value)) {
    axis = value === UNICODE.arrowLeft || value === UNICODE.arrowRight ? "horizontal" : "vertical";
  } else return null;

  const dirs = axis === "horizontal" ? [Direction.LEFT, Direction.RIGHT] : [Direction.UP, Direction.DOWN];
  const connected = dirs.filter((d) => connects(layer.get(p.add(d)), d.opposite()));
  if (connected.length !== 1) return null;
  return { tip: p, axis, bodyDir: connected[0]!, arrow: isArrow(value) ? value : null };
}

const BENDS = new Set([
  UNICODE.cornerTopLeft,
  UNICODE.cornerTopRight,
  UNICODE.cornerBottomRight,
  UNICODE.cornerBottomLeft,
]);

const straightCharFor = (d: Direction) => (d.isHorizontal() ? H : V);

/**
 * Walks from a tip toward the body and stops at the first bend (the pivot).
 * Returns the cells from the tip up to and including that corner.
 */
export function traceLineFromTip(
  layer: Layer,
  tip: Vector,
  bodyDir: Direction,
): { cells: Vector[]; anchor: Vector } {
  const cells: Vector[] = [tip];
  let current = tip.add(bodyDir);
  for (let steps = 0; steps < 1000; steps++) {
    const value = layer.get(current);
    if (value === straightCharFor(bodyDir)) {
      cells.push(current);
      current = current.add(bodyDir);
      continue;
    }
    if (value != null && BENDS.has(value) && connects(value, bodyDir.opposite())) {
      cells.push(current);
      return { cells, anchor: current };
    }
    break;
  }
  return { cells, anchor: cells[cells.length - 1]! };
}

const isVerticalBorder = (v: string | null) =>
  isBoxDrawing(v) && (connects(v, Direction.UP) || connects(v, Direction.DOWN));
const isHorizontalBorder = (v: string | null) =>
  isBoxDrawing(v) && (connects(v, Direction.LEFT) || connects(v, Direction.RIGHT));

const RAY_LIMIT = 400;

function ray(layer: Layer, from: Vector, d: Direction, pred: (v: string | null) => boolean): Vector | null {
  let p = from;
  for (let i = 0; i < RAY_LIMIT; i++) {
    p = p.add(d);
    if (pred(layer.get(p))) return p;
  }
  return null;
}

function verifyPerimeter(layer: Layer, box: Box): boolean {
  for (let x = box.left(); x <= box.right(); x++) {
    if (!isBoxDrawing(layer.get(new Vector(x, box.top())))) return false;
    if (!isBoxDrawing(layer.get(new Vector(x, box.bottom())))) return false;
  }
  for (let y = box.top(); y <= box.bottom(); y++) {
    if (!isBoxDrawing(layer.get(new Vector(box.left(), y)))) return false;
    if (!isBoxDrawing(layer.get(new Vector(box.right(), y)))) return false;
  }
  return true;
}

function boxFromSeed(layer: Layer, seed: Vector): Box | null {
  const left = ray(layer, seed, Direction.LEFT, isVerticalBorder);
  const right = ray(layer, seed, Direction.RIGHT, isVerticalBorder);
  const up = ray(layer, seed, Direction.UP, isHorizontalBorder);
  const down = ray(layer, seed, Direction.DOWN, isHorizontalBorder);
  if (!left || !right || !up || !down) return null;
  const box = new Box(new Vector(left.x, up.y), new Vector(right.x, down.y));
  if (box.right() - box.left() < 1 || box.bottom() - box.top() < 1) return null;
  return verifyPerimeter(layer, box) ? box : null;
}

const area = (b: Box) => (b.right() - b.left()) * (b.bottom() - b.top());

const DIAGONALS = [new Vector(-1, -1), new Vector(1, -1), new Vector(-1, 1), new Vector(1, 1)];
const COMPONENT_LIMIT = 4000;

function borderComponent(layer: Layer, start: Vector): Vector[] | null {
  const seen = new Set([start.key()]);
  const stack = [start];
  const cells: Vector[] = [];
  while (stack.length > 0) {
    const current = stack.pop()!;
    cells.push(current);
    if (cells.length > COMPONENT_LIMIT) return null;
    for (const d of Direction.ALL) {
      const next = current.add(d);
      if (!seen.has(next.key()) && isBoxDrawing(layer.get(next))) {
        seen.add(next.key());
        stack.push(next);
      }
    }
  }
  return cells;
}

/** Smallest closed box enclosing `p` (on its border or inside), or null. */
export function findBox(layer: Layer, p: Vector): Box | null {
  const candidates: Box[] = [];
  const seeds: Vector[] = [];
  if (!isBoxDrawing(layer.get(p))) {
    seeds.push(p);
  } else {
    for (const d of [...Direction.ALL, ...DIAGONALS]) {
      const n = p.add(d);
      if (!isBoxDrawing(layer.get(n))) seeds.push(n);
    }
  }
  for (const seed of seeds) {
    const box = boxFromSeed(layer, seed);
    if (box && box.contains(p)) candidates.push(box);
  }
  if (isBoxDrawing(layer.get(p))) {
    const component = borderComponent(layer, p);
    const bounds = component && boundingBox(component);
    if (
      bounds &&
      bounds.right() - bounds.left() >= 1 &&
      bounds.bottom() - bounds.top() >= 1 &&
      bounds.contains(p) &&
      verifyPerimeter(layer, bounds)
    ) {
      candidates.push(bounds);
    }
  }
  let best: Box | null = null;
  for (const b of candidates) if (!best || area(b) < area(best)) best = b;
  return best;
}

export function cellsInBox(layer: Layer, box: Box): Vector[] {
  const out: Vector[] = [];
  for (const [k, v] of layer.map) {
    if (isErase(v)) continue;
    const p = Vector.fromKey(k);
    if (box.contains(p)) out.push(p);
  }
  return out;
}

/** Translates `cells` by `delta`, snapping the surrounding structure. */
export function moveCells(committed: Layer, cells: readonly Vector[], delta: Vector): Layer {
  const layer = new Layer();
  for (const c of cells) layer.set(c, "");
  const protect = new Set<string>();
  for (const c of cells) {
    const v = committed.get(c);
    if (isContent(v)) {
      const t = c.add(delta);
      layer.set(t, v);
      protect.add(t.key());
    }
  }
  layer.setFrom(snap(layer, committed, protect));
  return layer;
}

export interface BoxAttachment {
  /** Border cell of the box the line attaches to. */
  anchor: Vector;
  /** Outward direction, away from the box. */
  out: Direction;
  /** Fixed endpoint the connector must still reach. */
  far: Vector;
  /** Connector cells between the box and `far`, corners included. */
  runCells: Vector[];
  /** The line ends in an arrow head pointing into the box. */
  arrowIntoBox: boolean;
}

function turnFrom(value: string, dir: Direction): Direction | null {
  for (const d of connections(value)) if (d !== dir.opposite()) return d;
  return null;
}

function perimeter(box: Box): [Vector, Direction][] {
  const out: [Vector, Direction][] = [];
  for (let x = box.left(); x <= box.right(); x++) {
    out.push([new Vector(x, box.top()), Direction.UP]);
    out.push([new Vector(x, box.bottom()), Direction.DOWN]);
  }
  for (let y = box.top(); y <= box.bottom(); y++) {
    out.push([new Vector(box.left(), y), Direction.LEFT]);
    out.push([new Vector(box.right(), y), Direction.RIGHT]);
  }
  return out;
}

/** Lines leaving `box`, each followed through corners to its terminal. */
export function traceBoxAttachments(layer: Layer, box: Box): BoxAttachment[] {
  const attachments: BoxAttachment[] = [];
  for (const [anchor, out] of perimeter(box)) {
    const first = anchor.add(out);
    const firstValue = layer.get(first);
    const straightOut = firstValue === straightCharFor(out) && connects(firstValue, out.opposite());
    const behind = first.add(out);
    const arrowIntoBox =
      firstValue === ARROW_FOR.get(out.opposite()) &&
      layer.get(behind) === straightCharFor(out) &&
      connects(layer.get(behind), out.opposite());
    if (!straightOut && !arrowIntoBox) continue;

    const runCells: Vector[] = [];
    let dir = out;
    let current = first;
    if (arrowIntoBox) {
      runCells.push(first);
      current = behind;
    }
    for (let steps = 0; steps < 1000; steps++) {
      const value = layer.get(current);
      if (value === straightCharFor(dir)) {
        runCells.push(current);
        current = current.add(dir);
        continue;
      }
      if (value != null && BENDS.has(value) && connects(value, dir.opposite())) {
        const next = turnFrom(value, dir);
        if (!next) break;
        runCells.push(current);
        dir = next;
        current = current.add(dir);
        continue;
      }
      break;
    }
    attachments.push({ anchor, out, far: current, runCells, arrowIntoBox });
  }
  return attachments;
}

function cornerFor(a: Direction, b: Direction): string {
  const h = a.isHorizontal() ? a : b;
  const v = a.isHorizontal() ? b : a;
  if (h === Direction.LEFT) return v === Direction.UP ? UNICODE.cornerBottomRight : UNICODE.cornerTopRight;
  return v === Direction.UP ? UNICODE.cornerBottomLeft : UNICODE.cornerTopLeft;
}

function drawStraight(layer: Layer, from: Vector, to: Vector, excludeEnd: boolean) {
  const step = new Vector(Math.sign(to.x - from.x), Math.sign(to.y - from.y));
  const ch = step.x !== 0 ? H : V;
  let p = from;
  while (true) {
    if (!(excludeEnd && p.equals(to))) layer.set(p, ch);
    if (p.equals(to)) break;
    p = p.add(step);
  }
}

function drawConnector(layer: Layer, from: Vector, to: Vector, out: Direction, farValue: string | null) {
  if (from.equals(to)) return;
  const horizontalFirst = out.isHorizontal();
  const bend = horizontalFirst ? new Vector(to.x, from.y) : new Vector(from.x, to.y);
  drawStraight(layer, from, bend, bend.equals(to));
  let approach = new Vector(out.x, out.y);
  if (!bend.equals(to)) {
    drawStraight(layer, bend, to, true);
    approach = new Vector(Math.sign(to.x - bend.x), Math.sign(to.y - bend.y));
  }
  if (!bend.equals(from) && !bend.equals(to)) {
    const next = horizontalFirst
      ? to.y > bend.y
        ? Direction.DOWN
        : Direction.UP
      : to.x > bend.x
        ? Direction.RIGHT
        : Direction.LEFT;
    layer.set(bend, cornerFor(out.opposite(), next));
  }
  if (isArrow(farValue)) {
    for (const [d, ch] of ARROW_FOR) {
      if (d.x === approach.x && d.y === approach.y) layer.set(to, ch);
    }
  }
}

/** Moves a box and its contents by `delta`, reflowing attached lines. */
export function moveBoxWithAttachments(
  committed: Layer,
  box: Box,
  attachments: readonly BoxAttachment[],
  delta: Vector,
): Layer {
  const layer = new Layer();
  const content = cellsInBox(committed, box);
  for (const c of content) layer.set(c, "");
  for (const a of attachments) for (const c of a.runCells) layer.set(c, "");
  for (const c of content) {
    const v = committed.get(c);
    if (isContent(v)) layer.set(c.add(delta), v);
  }
  for (const a of attachments) {
    const newAnchor = a.anchor.add(a.out).add(delta);
    drawConnector(layer, newAnchor, a.far, a.out, committed.get(a.far));
    if (a.arrowIntoBox) layer.set(newAnchor, ARROW_FOR.get(a.out.opposite())!);
  }
  const protect = new Set(content.map((c) => c.add(delta).key()));
  layer.setFrom(snap(layer, committed, protect));
  return layer;
}

File: src/core/entity.rs
//! Entity detection (words, line tips, boxes) and moves that keep attached
//! lines connected.
//!
//! Ported from ASCIIFlow (`client/draw/entity.ts`), MIT © Lewis Hemens.

use std::collections::HashSet;

use super::glyphs::{arrows, connections, connects, is_arrow, is_box_drawing, UNICODE};
use super::grid::{bounding_box, Bounds};
use super::layer::{is_erase, Layer, ERASE};
use super::snap::snap;
use super::vector::{Direction, Pos};

const H: char = UNICODE.line_horizontal;
const V: char = UNICODE.line_vertical;

pub fn is_content(c: char) -> bool {
    !is_erase(c)
}

pub fn is_text(c: char) -> bool {
    is_content(c) && !is_box_drawing(c)
}

/// Maximal horizontal run of text characters under `p`.
pub fn detect_word(layer: &Layer, p: Pos) -> Option<Vec<Pos>> {
    if !layer.get(p).is_some_and(is_text) {
        return None;
    }
    let mut left = p;
    let mut right = p;
    while layer.get(left.left()).is_some_and(is_text) {
        left = left.left();
    }
    while layer.get(right.right()).is_some_and(is_text) {
        right = right.right();
    }
    Some((left.x..=right.x).map(|x| Pos::new(x, p.y)).collect())
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Axis {
    Horizontal,
    Vertical,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct LineTip {
    pub tip: Pos,
    pub axis: Axis,
    /// Direction from the tip toward the rest of the line.
    pub body_dir: Direction,
    pub arrow: Option<char>,
}

/// The free end of a line or arrow under `p`, if any.
pub fn detect_line_tip(layer: &Layer, p: Pos) -> Option<LineTip> {
    let value = layer.get(p)?;
    let axis = match value {
        H => Axis::Horizontal,
        V => Axis::Vertical,
        c if is_arrow(c) => {
            if c == UNICODE.arrow_left || c == UNICODE.arrow_right {
                Axis::Horizontal
            } else {
                Axis::Vertical
            }
        }
        _ => return None,
    };
    let dirs = match axis {
        Axis::Horizontal => [Direction::Left, Direction::Right],
        Axis::Vertical => [Direction::Up, Direction::Down],
    };
    let connected: Vec<Direction> = dirs
        .into_iter()
        .filter(|d| {
            layer
                .get(p.add(d.delta()))
                .is_some_and(|n| connects(n, d.opposite()))
        })
        .collect();
    if connected.len() != 1 {
        return None;
    }
    Some(LineTip {
        tip: p,
        axis,
        body_dir: connected[0],
        arrow: is_arrow(value).then_some(value),
    })
}

fn is_bend(c: char) -> bool {
    c == UNICODE.corner_top_left
        || c == UNICODE.corner_top_right
        || c == UNICODE.corner_bottom_right
        || c == UNICODE.corner_bottom_left
}

fn straight_char_for(d: Direction) -> char {
    if d.is_horizontal() {
        H
    } else {
        V
    }
}

pub struct TipTrace {
    pub cells: Vec<Pos>,
    pub anchor: Pos,
}

/// Walks from a tip toward the body and stops at the first bend (the pivot).
/// Returns the cells from the tip up to and including that corner.
pub fn trace_line_from_tip(layer: &Layer, tip: Pos, body_dir: Direction) -> TipTrace {
    let mut cells = vec![tip];
    let mut current = tip.add(body_dir.delta());
    for _ in 0..1000 {
        let value = layer.get(current);
        if value == Some(straight_char_for(body_dir)) {
            cells.push(current);
            current = current.add(body_dir.delta());
            continue;
        }
        if let Some(v) = value {
            if is_bend(v) && connects(v, body_dir.opposite()) {
                cells.push(current);
                return TipTrace {
                    cells,
                    anchor: current,
                };
            }
        }
        break;
    }
    let anchor = *cells.last().expect("the tip is always present");
    TipTrace { cells, anchor }
}

fn is_vertical_border(c: Option<char>) -> bool {
    c.is_some_and(|v| {
        is_box_drawing(v) && (connects(v, Direction::Up) || connects(v, Direction::Down))
    })
}

fn is_horizontal_border(c: Option<char>) -> bool {
    c.is_some_and(|v| {
        is_box_drawing(v) && (connects(v, Direction::Left) || connects(v, Direction::Right))
    })
}

const RAY_LIMIT: usize = 400;

fn ray(layer: &Layer, from: Pos, d: Direction, pred: impl Fn(Option<char>) -> bool) -> Option<Pos> {
    let mut p = from;
    for _ in 0..RAY_LIMIT {
        p = p.add(d.delta());
        if pred(layer.get(p)) {
            return Some(p);
        }
    }
    None
}

fn verify_perimeter(layer: &Layer, b: Bounds) -> bool {
    for x in b.left()..=b.right() {
        if !layer.get(Pos::new(x, b.top())).is_some_and(is_box_drawing) {
            return false;
        }
        if !layer
            .get(Pos::new(x, b.bottom()))
            .is_some_and(is_box_drawing)
        {
            return false;
        }
    }
    for y in b.top()..=b.bottom() {
        if !layer.get(Pos::new(b.left(), y)).is_some_and(is_box_drawing) {
            return false;
        }
        if !layer
            .get(Pos::new(b.right(), y))
            .is_some_and(is_box_drawing)
        {
            return false;
        }
    }
    true
}

fn box_from_seed(layer: &Layer, seed: Pos) -> Option<Bounds> {
    let left = ray(layer, seed, Direction::Left, is_vertical_border)?;
    let right = ray(layer, seed, Direction::Right, is_vertical_border)?;
    let up = ray(layer, seed, Direction::Up, is_horizontal_border)?;
    let down = ray(layer, seed, Direction::Down, is_horizontal_border)?;
    let b = Bounds::new(Pos::new(left.x, up.y), Pos::new(right.x, down.y));
    if b.right() - b.left() < 1 || b.bottom() - b.top() < 1 {
        return None;
    }
    verify_perimeter(layer, b).then_some(b)
}

fn area(b: &Bounds) -> i32 {
    (b.right() - b.left()) * (b.bottom() - b.top())
}

const DIAGONALS: [Pos; 4] = [
    Pos::new(-1, -1),
    Pos::new(1, -1),
    Pos::new(-1, 1),
    Pos::new(1, 1),
];

const COMPONENT_LIMIT: usize = 4000;

fn border_component(layer: &Layer, start: Pos) -> Option<Vec<Pos>> {
    let mut seen: HashSet<Pos> = HashSet::from([start]);
    let mut stack = vec![start];
    let mut cells = Vec::new();
    while let Some(current) = stack.pop() {
        cells.push(current);
        if cells.len() > COMPONENT_LIMIT {
            return None;
        }
        for d in Direction::ALL {
            let next = current.add(d.delta());
            if !seen.contains(&next) && layer.get(next).is_some_and(is_box_drawing) {
                seen.insert(next);
                stack.push(next);
            }
        }
    }
    Some(cells)
}

/// Smallest closed box enclosing `p` (on its border or inside), or `None`.
pub fn find_box(layer: &Layer, p: Pos) -> Option<Bounds> {
    let mut candidates: Vec<Bounds> = Vec::new();
    let mut seeds: Vec<Pos> = Vec::new();
    if !layer.get(p).is_some_and(is_box_drawing) {
        seeds.push(p);
    } else {
        for d in Direction::ALL
            .into_iter()
            .map(Direction::delta)
            .chain(DIAGONALS)
        {
            let n = p.add(d);
            if !layer.get(n).is_some_and(is_box_drawing) {
                seeds.push(n);
            }
        }
    }
    for seed in seeds {
        if let Some(b) = box_from_seed(layer, seed) {
            if b.contains(p) {
                candidates.push(b);
            }
        }
    }
    if layer.get(p).is_some_and(is_box_drawing) {
        let bounds = border_component(layer, p).and_then(bounding_box);
        if let Some(b) = bounds {
            if b.right() - b.left() >= 1
                && b.bottom() - b.top() >= 1
                && b.contains(p)
                && verify_perimeter(layer, b)
            {
                candidates.push(b);
            }
        }
    }
    candidates
        .into_iter()
        .fold(None::<Bounds>, |best, b| match best {
            Some(prev) if area(&prev) <= area(&b) => Some(prev),
            _ => Some(b),
        })
}

pub fn cells_in_box(layer: &Layer, b: Bounds) -> Vec<Pos> {
    let mut out = Vec::new();
    for (p, v) in layer.entries() {
        if is_erase(v) {
            continue;
        }
        if b.contains(p) {
            out.push(p);
        }
    }
    out
}

/// Translates `cells` by `delta`, snapping the surrounding structure.
pub fn move_cells(committed: &Layer, cells: &[Pos], delta: Pos) -> Layer {
    let mut layer = Layer::new();
    for &c in cells {
        layer.set(c, ERASE);
    }
    let mut protect: HashSet<Pos> = HashSet::new();
    for &c in cells {
        if let Some(v) = committed.get(c) {
            if is_content(v) {
                let t = c.add(delta);
                layer.set(t, v);
                protect.insert(t);
            }
        }
    }
    layer.set_from(&snap(&layer, committed, &protect));
    layer
}

/// A line leaving a box: where it attaches, where it must still reach, and the
/// cells in between.
pub struct BoxAttachment {
    /// Border cell of the box the line attaches to.
    pub anchor: Pos,
    /// Outward direction, away from the box.
    pub out: Direction,
    /// Fixed endpoint the connector must still reach.
    pub far: Pos,
    /// Connector cells between the box and `far`, corners included.
    pub run_cells: Vec<Pos>,
    /// The line ends in an arrow head pointing into the box.
    pub arrow_into_box: bool,
}

fn turn_from(value: char, dir: Direction) -> Option<Direction> {
    connections(value)
        .into_iter()
        .find(|d| *d != dir.opposite())
}

fn perimeter(b: Bounds) -> Vec<(Pos, Direction)> {
    let mut out = Vec::new();
    for x in b.left()..=b.right() {
        out.push((Pos::new(x, b.top()), Direction::Up));
        out.push((Pos::new(x, b.bottom()), Direction::Down));
    }
    for y in b.top()..=b.bottom() {
        out.push((Pos::new(b.left(), y), Direction::Left));
        out.push((Pos::new(b.right(), y), Direction::Right));
    }
    out
}

/// Lines leaving `box`, each followed through corners to its terminal.
pub fn trace_box_attachments(layer: &Layer, b: Bounds) -> Vec<BoxAttachment> {
    let mut attachments = Vec::new();
    for (anchor, out) in perimeter(b) {
        let first = anchor.add(out.delta());
        let first_value = layer.get(first);
        let straight_out = first_value == Some(straight_char_for(out))
            && first_value.is_some_and(|v| connects(v, out.opposite()));
        let behind = first.add(out.delta());
        let arrow_into_box = first_value == Some(super::glyphs::arrow_for(out.opposite()))
            && layer.get(behind) == Some(straight_char_for(out))
            && layer
                .get(behind)
                .is_some_and(|v| connects(v, out.opposite()));
        if !straight_out && !arrow_into_box {
            continue;
        }

        let mut run_cells = Vec::new();
        let mut dir = out;
        let mut current = first;
        if arrow_into_box {
            run_cells.push(first);
            current = behind;
        }
        for _ in 0..1000 {
            let value = layer.get(current);
            if value == Some(straight_char_for(dir)) {
                run_cells.push(current);
                current = current.add(dir.delta());
                continue;
            }
            if let Some(v) = value {
                if is_bend(v) && connects(v, dir.opposite()) {
                    let Some(next) = turn_from(v, dir) else { break };
                    run_cells.push(current);
                    dir = next;
                    current = current.add(dir.delta());
                    continue;
                }
            }
            break;
        }
        attachments.push(BoxAttachment {
            anchor,
            out,
            far: current,
            run_cells,
            arrow_into_box,
        });
    }
    attachments
}

fn corner_for(a: Direction, b: Direction) -> char {
    let h = if a.is_horizontal() { a } else { b };
    let v = if a.is_horizontal() { b } else { a };
    if h == Direction::Left {
        if v == Direction::Up {
            UNICODE.corner_bottom_right
        } else {
            UNICODE.corner_top_right
        }
    } else if v == Direction::Up {
        UNICODE.corner_bottom_left
    } else {
        UNICODE.corner_top_left
    }
}

fn draw_straight(layer: &mut Layer, from: Pos, to: Pos, exclude_end: bool) {
    let step = Pos::new((to.x - from.x).signum(), (to.y - from.y).signum());
    let ch = if step.x != 0 { H } else { V };
    let mut p = from;
    loop {
        if !(exclude_end && p == to) {
            layer.set(p, ch);
        }
        if p == to {
            break;
        }
        p = p.add(step);
    }
}

fn draw_connector(layer: &mut Layer, from: Pos, to: Pos, out: Direction, far_value: Option<char>) {
    if from == to {
        return;
    }
    let horizontal_first = out.is_horizontal();
    let bend = if horizontal_first {
        Pos::new(to.x, from.y)
    } else {
        Pos::new(from.x, to.y)
    };
    draw_straight(layer, from, bend, bend == to);
    let mut approach = out.delta();
    if bend != to {
        draw_straight(layer, bend, to, true);
        approach = Pos::new((to.x - bend.x).signum(), (to.y - bend.y).signum());
    }
    if bend != from && bend != to {
        let next = if horizontal_first {
            if to.y > bend.y {
                Direction::Down
            } else {
                Direction::Up
            }
        } else if to.x > bend.x {
            Direction::Right
        } else {
            Direction::Left
        };
        layer.set(bend, corner_for(out.opposite(), next));
    }
    if far_value.is_some_and(is_arrow) {
        for (d, ch) in arrows() {
            if d.delta() == approach {
                layer.set(to, ch);
            }
        }
    }
}

/// Moves a box and its contents by `delta`, reflowing attached lines.
pub fn move_box_with_attachments(
    committed: &Layer,
    b: Bounds,
    attachments: &[BoxAttachment],
    delta: Pos,
) -> Layer {
    let mut layer = Layer::new();
    let content = cells_in_box(committed, b);
    for &c in &content {
        layer.set(c, ERASE);
    }
    for a in attachments {
        for &c in &a.run_cells {
            layer.set(c, ERASE);
        }
    }
    for &c in &content {
        if let Some(v) = committed.get(c) {
            if is_content(v) {
                layer.set(c.add(delta), v);
            }
        }
    }
    for a in attachments {
        let new_anchor = a.anchor.add(a.out.delta()).add(delta);
        draw_connector(&mut layer, new_anchor, a.far, a.out, committed.get(a.far));
        if a.arrow_into_box {
            layer.set(new_anchor, super::glyphs::arrow_for(a.out.opposite()));
        }
    }
    let protect: HashSet<Pos> = content.iter().map(|c| c.add(delta)).collect();
    layer.set_from(&snap(&layer, committed, &protect));
    layer
}
